llvm-external-projects/iree-dialects/test/python/iree_pydm/importer/function_def.py

```diff
@@ -0,0 +1,21 @@
+# RUN: %PYTHON %s | iree-dialects-opt | FileCheck --enable-var-scope --dump-input-filter=all %s
+
+from typing import List
+from mlir.dialects.iree_pydm.importer.test_util import *
+
+
+# CHECK-LABEL: iree_pydm.func @fully_typed_with_return
+# CHECK-SAME: (%arg0: !iree_pydm.integer) -> (!iree_pydm.exception_result, !iree_pydm.integer)
+# CHECK-SAME: attributes {arg_names = ["a"], cell_vars = [], free_vars = ["a"]}
```

Comment:
could you put the arg names as arg attrs instead? that would make it easier when doing lowering to ensure they are preserved consistently and as arguments are removed the op stays valid (as otherwise this is a parallel data structure that would need to be updated)

the FunctionLike trait provides a bunch of support for this (automatically handles inserting/erasing, has accessors, etc): https://github.com/llvm/llvm-project/blob/8066f22c4663d9ee6c763d9108c89448e5c19848/mlir/include/mlir/IR/FunctionSupport.h#L377

Reply:
Left a comment: `// TODO: Move arg attributes to arguments, and generally rework the free_vars,
// cell_vars attributes (with better modeling, they may not be needed at all).`

The Python FunctionLike support doesn't have any of the niceties yet and I need to work that out. Will do shortly as it will be pretty obvious what would work better when actually writing transformations on this.